pkg/util/validation/limits.go

```diff
@@ -0,0 +1,226 @@
+type Limits struct {
+	// Distributor enforced limits.
+	IngestionRate          float64       `yaml:"ingestion_rate"`
+	IngestionBurstSize     float64       `yaml:"ingestion_burst_size"`
+	MaxLabelNameLength     int           `yaml:"max_label_name_length"`
+	MaxLabelValueLength    int           `yaml:"max_label_value_length"`
+	MaxLabelNamesPerSeries int           `yaml:"max_label_names_per_series"`
+	RejectOldSamples       bool          `yaml:"reject_old_samples"`
+	RejectOldSamplesMaxAge time.Duration `yaml:"reject_old_samples_max_age"`
+	CreationGracePeriod    time.Duration `yaml:"creation_grace_period"`
+	EnforceMetricName      bool          `yaml:"enforce_metric_name"`
+
+	// Ingester enforced limits.
+	MaxStreamsPerUser int `yaml:"max_streams_per_user"`
+
+	// Querier enforced limits.
+	MaxChunksPerQuery          int           `yaml:"max_chunks_per_query"`
+	MaxQueryLength             time.Duration `yaml:"max_query_length"`
+	MaxQueryParallelism        int           `yaml:"max_query_parallelism"`
+	CardinalityLimit           int           `yaml:"cardinality_limit"`
+	MaxStreamsMatchersPerQuery int           `yaml:"max_streams_matchers_per_query"`
+
+	// Config for overrides, convenient if it goes here.
+	PerTenantOverrideConfig string        `yaml:"per_tenant_override_config"`
+	PerTenantOverridePeriod time.Duration `yaml:"per_tenant_override_period"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (l *Limits) RegisterFlags(f *flag.FlagSet) {
+	f.Float64Var(&l.IngestionRate, "distributor.ingestion-rate-limit", 1, "Per-user ingestion rate limit in sample size per second.")
+	f.Float64Var(&l.IngestionBurstSize, "distributor.ingestion-burst-size", 1.5, "Per-user allowed ingestion burst size (in sample size). Warning, very high limits will be reset every -distributor.limiter-reload-period.")
+	f.IntVar(&l.MaxLabelNameLength, "validation.max-length-label-name", 1024, "Maximum length accepted for label names")
+	f.IntVar(&l.MaxLabelValueLength, "validation.max-length-label-value", 2048, "Maximum length accepted for label value. This setting also applies to the metric name")
+	f.IntVar(&l.MaxLabelNamesPerSeries, "validation.max-label-names-per-series", 30, "Maximum number of label names per series.")
+	f.BoolVar(&l.RejectOldSamples, "validation.reject-old-samples", false, "Reject old samples.")
+	f.DurationVar(&l.RejectOldSamplesMaxAge, "validation.reject-old-samples.max-age", 14*24*time.Hour, "Maximum accepted sample age before rejecting.")
+	f.DurationVar(&l.CreationGracePeriod, "validation.create-grace-period", 10*time.Minute, "Duration which table will be created/deleted before/after it's needed; we won't accept sample from before this time.")
+	f.BoolVar(&l.EnforceMetricName, "validation.enforce-metric-name", true, "Enforce every sample has a metric name.")
+
+	f.IntVar(&l.MaxStreamsPerUser, "ingester.max-streams-per-user", 5000000, "Maximum number of active streams per user.")
```

Comment:
This default seems much too high to me?  I would think we would want to default to something much lower.  I pulled some stats from our current Loki:

ops-tools has about 30,000 active streams, one of our trial customers on us-west1 has 188,000! 
Everyone else on us-west1 is less than 1,000.

I think we should start the limit 10,000? (and we should probably use the 10e3 notation to be consistent with other limits)

@gouthamve @tomwilkie WDYT? you have a good gut feeling on what we should set default cardinality limit for log streams?

Reply:
:+1:  for lowering the max streams limit. My intuition on this is that 5000000 is a more appropriate limit for Cortex because the metric name is an extra factor in how many series there are. Loki doesn't have that and so I feel like overall there's going to be less streams in general. 

10e3 sounds like a good starting point to me.